repair/row_level.cc

```diff
@@ -1870,37 +1882,58 @@ class repair_meta {
     future<> put_row_diff_with_rpc_stream(
             repair_hash_set set_diff,
             needs_all_rows_t needs_all_rows,
-            gms::inet_address remote_node, unsigned node_idx) {
-        if (!set_diff.empty()) {
-            if (remote_node == _myip) {
-                return make_ready_future<>();
-            }
-            size_t sz = set_diff.size();
-            return get_row_diff(std::move(set_diff), needs_all_rows).then([this, remote_node, node_idx, sz] (std::list<repair_row> row_diff) {
-                if (row_diff.size() != sz) {
-                    rlogger.warn("Hash conflict detected, keyspace={}, table={}, range={}, row_diff.size={}, set_diff.size={}. It is recommended to compact the table and rerun repair for the range.",
-                            _schema->ks_name(), _schema->cf_name(), _range, row_diff.size(), sz);
+            gms::inet_address remote_node, unsigned node_idx,
+            const locator::effective_replication_map& erm, bool small_table_optimization) {
+        if (set_diff.empty()) {
+            co_return;
+        }
+        if (remote_node == _myip) {
+            co_return;
+        }
+        size_t sz = set_diff.size();
+
+        std::list<repair_row> row_diff = co_await get_row_diff(std::move(set_diff), needs_all_rows);
+        if (row_diff.size() != sz) {
+            rlogger.warn("Hash conflict detected, keyspace={}, table={}, range={}, row_diff.size={}, set_diff.size={}. It is recommended to compact the table and rerun repair for the range.",
+                    _schema->ks_name(), _schema->cf_name(), _range, row_diff.size(), sz);
+        }
+        if (small_table_optimization) {
+            auto& strat = erm.get_replication_strategy();
+            auto& tm = erm.get_token_metadata();
+            std::list<repair_row> tmp;
+            for (auto& row : row_diff) {
+                repair_row r = std::move(row);
+                const auto& dk = r.get_dk_with_hash()->dk;
+                auto eps = co_await strat.calculate_natural_endpoints(dk.token(), tm);
```

Comment:
Yes. We could add a boost factor when calculating the size of the table when the avg size is small as an optimization. Since the totally amount of data is small, the check overhead above is not that important. For instance, the worse case 1B per partition , 100MB size means 100M partitions. Checking 100M partitions ownership is not that bad.

Reply:
ditto re: `calculate_natural_endpoints`.
here too, I don't see why we call the strategy method rather `erm.get_natural_endpoints`

Cc @tgrabiec